06-security.md

```diff
@@ -1 +1,451 @@
+    CsrfToken csrf = (CsrfToken) request.getAttribute(CsrfToken.class.getName());
+    if (csrf != null) {
+      Cookie cookie = WebUtils.getCookie(request, "XSRF-TOKEN");
+      String token = csrf.getToken();
+      if (cookie==null || token!=null && !token.equals(cookie.getValue())) {
+        cookie = new Cookie("XSRF-TOKEN", token);
+        cookie.setSecure(request.isSecure());
+        cookie.setPath(StringUtils.isEmpty(request.getContextPath()) ? "/" : request.getContextPath());
+        response.addCookie(cookie);
+      }
+    }
+    filterChain.doFilter(request, response);
+  }
+}
+```
+
+Also the filter must be added to the Spring Security Filter Chain:
+
+```java
+protected void configure(HttpSecurity http) throws Exception {
+    http.
+      ...
+      .addFilterAfter(new XsrfHeaderFilter(), CsrfFilter.class);
+}
+```
+
+#### Verifying the XSRF token	
+
+For each modifying request (such as POST) Spring Security expects the XSRF token. The defaults don't match those of AngularJS which sends the token as a `X-XSRF-TOKEN` header (Spring Security expects an `X-CSRF-TOKEN` header) so some slight configuration is required.  
+
+```java
+@Override
+protected void configure(HttpSecurity http) throws Exception {
+  http.
+    ...
+    .csrf().csrfTokenRepository(csrfTokenRepository());
+}
+
+private CsrfTokenRepository csrfTokenRepository() {
+  HttpSessionCsrfTokenRepository repository = new HttpSessionCsrfTokenRepository();
+  repository.setHeaderName("X-XSRF-TOKEN");
+  return repository;
+}
+```
+
+### Validating Strings
+
+An example `BasicStringValidator` and `@BasicString` annotation have been provided which will reject any control characters.  
+For more information on writing custom hibernate-validator validations read the [reference documentation] (https://docs.jboss.org/hibernate/validator/5.0/reference/en-US/html/validator-customconstraints.html). 
+
+# Further reading
+
+* [OWASP REST Security Cheat sheet](https://www.owasp.org/index.php/REST_Security_Cheat_Sheet)
+* [CORS with Spring MVC](http://dontpanic.42.nl/2015/04/cors-with-spring-mvc.html)
+* [Angular JS and Spring Security](https://spring.io/guides/tutorials/spring-security-and-angular-js/)
+* [Angular JS and Spring Security Part II](https://spring.io/blog/2015/01/12/the-login-page-angular-js-and-spring-security-part-ii)
+* [Spring Security CSRF Reference](https://docs.spring.io/spring-security/site/docs/current/reference/html/csrf.html)
+* [How to Hack an API and get away with it](http://blog.smartbear.com/readyapi/api-security-testing-how-to-hack-an-api-and-get-away-with-it-part-1-of-3/)
+* [Understanding HTTP Strict Transport Security](http://www.troyhunt.com/2015/06/understanding-http-strict-transport.html)
+* [Secure REST Services using Spring Security](https://dzone.com/articles/secure-rest-services-using)
+* [OAuth 2.0 v.s. Json Web Tokens](http://www.seedbox.com/en/blog/2015/06/05/oauth-2-vs-json-web-tokens-comment-securiser-un-api/)
```

Comment:
Not sure if you already have it, but it might be worthwhile adding an example file for your Instant Developer Experience, setting up users if connection is made with an external authentication store, such as Crowd:

```
@Configuration
@Profile("local-crowd")
@EnableWebSecurity
public class LocalSecurityConfig extends AbstractSecurityConfig {

    @Override
    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
        auth
            .inMemoryAuthentication()
                .withUser("dev").password("xxx").roles("snip-users", "kauw-users", "kauw-manager")
            .and()
                .withUser("user").password("xxx").roles("user", "snip-users", "kauw-users", "user")
            .and()
                .withUser("pl").password("xxx").roles("user", "pl", "trusted")
            .and()
                .withUser("hr").password("xxx").roles("user", "hr", "trusted")
            .and()
                .withUser("mt").password("xxx").roles("user", "mt", "trusted", "management")
            .and()
                .withUser("office").password("xxx").roles("user", "office", "trusted", "management", "backoffice")
            .and()
                .withUser("control").password("xxx").roles("user", "control", "trusted", "management", "backoffice")
            .and()
                .withUser("admin").password("xxx").roles("user", "admin", "trusted", "management", "backoffice")
            .and()
                .withUser("person").password("xxx").roles("user");
    }
}
```